dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdRegistry.java

```diff
@@ -0,0 +1,344 @@
+            clientBuilder.user(byteSequence(registryProperties.getUser()));
+            clientBuilder.password(byteSequence(registryProperties.getPassword()));
+        }
+        if (!Strings.isNullOrEmpty(registryProperties.getLoadBalancerPolicy())) {
+            clientBuilder.loadBalancerPolicy(registryProperties.getLoadBalancerPolicy());
+        }
+        if (!Strings.isNullOrEmpty(registryProperties.getAuthority())) {
+            clientBuilder.authority(registryProperties.getAuthority());
+        }
+        client = clientBuilder.build();
+        LOGGER.info("Started Etcd Registry...");
+        etcdConnectionStateListener = new EtcdConnectionStateListener(client);
+    }
+
+    /**
+     * Start the etcd Connection stateListeer
+     */
+    @PostConstruct
+    public void start() {
+        LOGGER.info("Starting Etcd ConnectionListener...");
+        etcdConnectionStateListener.start();
+        LOGGER.info("Started Etcd ConnectionListener...");
+    }
+
+    /**
+     *
+     * @param path The prefix of the key being listened to
+     * @param listener
+     * @return if subcribe Returns true if no exception was thrown
+     */
+    @Override
+    public boolean subscribe(String path, SubscribeListener listener) {
+        try {
+            ByteSequence watchKey = byteSequence(path);
+            WatchOption watchOption = WatchOption.newBuilder().isPrefix(true).build();
+            watcherMap.computeIfAbsent(path, $ -> client.getWatchClient().watch(watchKey, watchOption,watchResponse -> {
+                for (WatchEvent event : watchResponse.getEvents()) {
+                    listener.notify(new EventAdaptor(event, path));
+                }
+            }));
+        } catch (Exception e) {
+            throw new RegistryException("Failed to subscribe listener for key: " + path, e);
+        }
+        return true;
+    }
+
+    /**
+     * @throws throws an exception if the unsubscribe path does not exist
+     * @param path The prefix of the key being listened to
+     */
+    @Override
+    public void unsubscribe(String path) {
+        try {
+            watcherMap.get(path).close();
+            watcherMap.remove(path);
+        } catch (Exception e) {
+            throw new RegistryException("Failed to unsubscribe listener for key: " + path, e);
+        }
+    }
+
+    @Override
+    public void addConnectionStateListener(ConnectionListener listener) {
+        etcdConnectionStateListener.addConnectionListener(listener);
+    }
+
+    /**
+     *
+     * @param key
+     * @return Returns the value corresponding to the key
+     * @throws throws an exception if the key does not exist
+     */
+    @Override
+    public String get(String key) {
+        try {
+            List<KeyValue> keyValues = client.getKVClient().get(byteSequence(key)).get().getKvs();
+            return keyValues.iterator().next().getValue().toString(StandardCharsets.UTF_8);
+        } catch (InterruptedException | ExecutionException e) {
```

Comment:
If throw ExecutionException, should we call `Thread.currentThread().interrupt()`? It's better to handle separately, WDYT?

Reply:
I only marked one place